### PR TITLE
Add Latex Renderer

### LINKED ISSUE #206: Latex Renderer
To include Perceval circuits in latex paper, we only have the option to generate svg through svg renderer  (`perceval/rendering/canvas/svg_canvas.py`) or png through mplot renderer (`perceval/rendering/canvas/mplot_canvas.py`). The renderer are automatically selected or can be specified with `output_format=Format.TEXT|LATEX|HTML` option. For circuit, LATEX format is not supported:

in `perceval/rendering/circuit/renderer.py`:

```
    if output_format == Format.LATEX:
        raise NotImplementedError("Latex format is not supported for circuit rendering")
```

**Describe the solution you'd like**
Having a LaTeX renderer would enable integration of higher quality circuits in scientific papers - we can integrate svg inside LaTeX using Inkscape, but it does not give the control of the generated graph for further LaTeX editing.
A solution could be to generate TikZ code that will then be fully editable within latex - for instance changing fontsize, etc...

### PR DESCRIPTION
**Pull request by king-p3nguin**
# Summary
- Resolve #206

## Details
- Running `pcvl.pdisplay(circuit, output_format=pcvl.rendering.Format.LATEX)` will output a series of LaTeX commands.

## Example
```python
import perceval as pcvl
import perceval.components.unitary_components as comp

mzi = (pcvl.Circuit(m=2, name="mzi")
       .add((0, 1), comp.BS())
       .add(0, comp.PS(pcvl.Parameter("phi1")))
       .add((0, 1), comp.BS())
       .add(0, comp.PS(pcvl.Parameter("phi2"))))

pcvl.pdisplay(mzi, output_format=pcvl.rendering.Format.LATEX)
```
Output:
```latex
\documentclass{standalone}
\usepackage{tikz}
\definecolor{darkred}{rgb}{0.55, 0.0, 0.0}
\definecolor{thistle}{rgb}{0.85, 0.75, 0.85}
\begin{document}
\begin{tikzpicture}[scale=1.5,x=1pt,y=1pt]
\draw[color=darkred,line width=3,line join=miter,fill=none] (10,-25) -- (25,-25);
\draw[color=darkred,line width=3,line join=miter,fill=none] (10,-75) -- (25,-75);
\draw[color=darkred,line width=3] (25,-25) -- (53,-25) -- (72,-44);
\draw[color=darkred,line width=3] (78,-44) -- (97,-25) -- (125,-25);
\draw[color=darkred,line width=3] (25,-75) -- (53,-75) -- (72,-56);
\draw[color=darkred,line width=3] (78,-56) -- (97,-75) -- (125,-75);
\draw[color=black,line width=1,fill=black] (50,-43) -- (100,-43) -- (100,-57) -- (50,-57) -- cycle;
\node[anchor=mid,font = {\fontsize{7pt}{0}}] at (75,-85) {};
\node[anchor=mid,font = {\fontsize{7pt}{0}}] at (75,-26) {};
\draw[color=black,line width=1,fill=lightgray] (50,-43) -- (100,-43) -- (100,-47) -- (50,-47) -- cycle;
\draw[color=black,line width=1,fill=thistle] (93,-50) -- (103,-50) -- (103,-60) -- (93,-60) -- cycle;
\node[anchor=mid,font = {\fontsize{6pt}{0}}] at (98,-57) {Rx};
\draw[color=darkred,line width=3] (125,-25) -- (175,-25);
\draw[color=black,line width=1,fill=gray] (130,-40) -- (139,-40) -- (153,-10) -- (144,-10) -- (130,-40) -- (139,-40) -- cycle;
\node[anchor=west,font = {\fontsize{7pt}{0}}] at (147,-38) {Φ=phi1};
\draw[color=darkred,line width=3] (125,-75) -- (175,-75);
\draw[color=darkred,line width=3] (175,-25) -- (203,-25) -- (222,-44);
\draw[color=darkred,line width=3] (228,-44) -- (247,-25) -- (275,-25);
\draw[color=darkred,line width=3] (175,-75) -- (203,-75) -- (222,-56);
\draw[color=darkred,line width=3] (228,-56) -- (247,-75) -- (275,-75);
\draw[color=black,line width=1,fill=black] (200,-43) -- (250,-43) -- (250,-57) -- (200,-57) -- cycle;
\node[anchor=mid,font = {\fontsize{7pt}{0}}] at (225,-85) {};
\node[anchor=mid,font = {\fontsize{7pt}{0}}] at (225,-26) {};
\draw[color=black,line width=1,fill=lightgray] (200,-43) -- (250,-43) -- (250,-47) -- (200,-47) -- cycle;
\draw[color=black,line width=1,fill=thistle] (243,-50) -- (253,-50) -- (253,-60) -- (243,-60) -- cycle;
\node[anchor=mid,font = {\fontsize{6pt}{0}}] at (248,-57) {Rx};
\draw[color=darkred,line width=3] (275,-25) -- (325,-25);
\draw[color=black,line width=1,fill=gray] (280,-40) -- (289,-40) -- (303,-10) -- (294,-10) -- (280,-40) -- (289,-40) -- cycle;
\node[anchor=west,font = {\fontsize{7pt}{0}}] at (297,-38) {Φ=phi2};
\draw[color=darkred,line width=3] (275,-75) -- (325,-75);
\draw[color=darkred,line width=3,line join=miter,fill=none] (325,-25) -- (340,-25);
\draw[color=darkred,line width=3,line join=miter,fill=none] (325,-75) -- (340,-75);
\node[anchor=east,font = {\fontsize{6pt}{0}}] at (350,-28) {0};
\node[anchor=east,font = {\fontsize{6pt}{0}}] at (350,-78) {1};
\node[anchor=west,font = {\fontsize{6pt}{0}}] at (0,-28) {0};
\node[anchor=west,font = {\fontsize{6pt}{0}}] at (0,-78) {1};
\end{tikzpicture}
\end{document}
```
Run 
```bash
$ lualatex file.tex
```
to create pdf file.

codetest

### CODE REVIEW

File: perceval/rendering/canvas/latex_canvas.py [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+    "lime": True,
+    "olive": True,
+    "orange": True,
+    "pink": True,
+    "purple": True,
+    "teal": True,
+    "violet": True,
+    "darkred": "\\definecolor{darkred}{rgb}{0.55, 0.0, 0.0}",
+    "thistle": "\\definecolor{thistle}{rgb}{0.85, 0.75, 0.85}",
+}
```
Comment (ericbrts) [inline perceval/rendering/canvas/latex_canvas.py:54]: There are colors we use that are missing from this dictionary.
For instance, "lightpink" and "lightblue" are missing (maybe others).

You can check this out with the following code:
```python
import perceval as pcvl

cnot = pcvl.catalog["postprocessed cnot"].as_processor().build()
pcvl.pdisplay(cnot, output_format=pcvl.Format.LATEX, recursive=False)
pcvl.pdisplay(cnot, output_format=pcvl.Format.LATEX, recursive=True)
```

File: perceval/rendering/canvas/latex_canvas.py [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+        ]
+        self._draws = []
+        self._postscrip = [
+            "\\end{tikzpicture}",
+            "\\end{document}",
+        ]
+        self._color_check_list = tikz_implemented_colors.copy()
+
+    def _check_color_is_implemented(self, color):
+        if self._color_check_list[color] is None:
```
Comment (ericbrts) [inline perceval/rendering/canvas/latex_canvas.py:76]: I think you meant `if color not in self._color_check_list:`

**Review by ericbrts — CHANGES_REQUESTED** [review]
Hello,
I'm Eric, one of Perceval maintainer. I've submitted several change requests to your pull request. I can help if needed.
In any case, thanks for your contribution that we're willing to include in Perceval!
Best wishes,
Eric

**Comment by ericbrts** [comment]
Also, I've seen that our current system tries to output UTF-8 greek letters (and this is not supported by LaTex). It has to be converted to a LaTex math notation (e.g. Θ must be changed to $\theta$ in LaTex code).

**Comment by ericbrts** [comment]
Hello @king-p3nguin ,
I still have issues with the output LaTex code (sorry!) :
image
It seems multiline math expression is not working as expected (actually we do not require it to be multiline, but we need it to work properly).

See the compiled output:
image
I've circled the issues in red.
* The multiline output is not working
* A smaller issue is that some shapes are not correctly filled with their color.

Do you think you can still work on these topics a bit?
Please let me know,
Best wishes,
Eric

**Comment by king-p3nguin** [comment]
Hi @ericbrts, I was able to fix both of the issues.
- Used `latexcodec` to convert UTF-8 characters to LaTeX style
- Support line breaking
- Fixed herald rendering

Here is the rendered image:
image

**Review by ericbrts — APPROVED** [review]

**Comment by ericbrts** [comment]
Hello @king-p3nguin ,

Thanks for your work and all of your latest fixes. I've just tested your code and everything is fine!
I'm merging now.

Cheers,
Eric

**Pull request merged**